### ISSUE TITLE
Catalog import open PR doesn't support GitHub Enterprise yet

### ISSUE THREAD

**Issue by adamdmharvey** [backstage/backstage#3541]
## Expected Behavior

When submitting the URL of a GitHub Enterprise repo to the register component screen, the pull request is opened on the GHE instance.

## Current Behavior

GitHub.com is hardcoded and so won't open PRs on GHE. User always receives HTTP 401 unauthorized.

Octokit library defaults to GitHub.com when not otherwise specified:

https://github.com/backstage/backstage/blob/257619c1ee9c49454fc32b6d7d2b07283ea8d738/plugins/catalog-import/src/api/CatalogImportClient.ts#L101-L103

Also hardcoded here:

https://github.com/backstage/backstage/blob/257619c1ee9c49454fc32b6d7d2b07283ea8d738/plugins/catalog-import/src/api/CatalogImportClient.ts#L180-L183

## Possible Solution

The OAuth provider objects don't seem to maintain the enterprise URL separately today, so I think if we query the config and get `enterpriseInstanceUrl` it could be used when instantiating the Octokit object.

Roughly this code:

https://github.com/backstage/backstage/blob/257619c1ee9c49454fc32b6d7d2b07283ea8d738/plugins/catalog-import/src/api/CatalogImportClient.ts#L90-L103

Would maybe need something along these lines though I wasn't able to get the plumbing from this component to the config working properly:

```
    const token = await this.githubAuthApi.getAccessToken(['repo']); 
    const configApi = useApi(configApiRef);
    const enterpriseInstanceUrl = configApi.getOptionalString('auth.providers.github.enterpriseInstanceUrl'); 
    const octo = new Octokit({
      auth: token,
      baseUrl: enterpriseInstanceUrl,
    });
```

The Octokit object just takes the `baseUrl` as a param.

## Steps to Reproduce

1. Register a component manually (e.g., http://localhost:3000/catalog-import )
2. Enter a GHE instance URL (even a fake one) in the form `https://ghe.mycompany.com/org/repo`, click Next
3. The review step CORRECTLY shows the target URL, though note the annotation is incorrect - it uses only the last part of the host and .com, in this example will only show "mycompany.com" when it should likely be hardcoded to github.com per the slug
    ![image](https://user-images.githubusercontent.com/33203301/100960142-b85e1880-34ed-11eb-8222-1835e071dab3.png)
4. Click Next, receive HTTP 401 from GitHub.com
    ![image](https://user-images.githubusercontent.com/33203301/100960057-877de380-34ed-11eb-9987-2a68b0a61cf7.png)

## Your Environment

- NodeJS Version (v12): v12.16.2
- Operating System and Version (e.g. Ubuntu 14.04): Windows 10 1909 WSL2 Ubuntu 18.04
- Browser Information: Microsoft Edge (Chrome engine 87)

**Comment by O5ten** [comment]
#3613 closes this issue.

**Comment by benjdlambert** [comment]
@adamdmharvey can you try with the merged work and then close this issue if it's all solved?

**Comment by adamdmharvey** [comment]
Roger that, opening the PR works now! Note, I snuck in two reports I guess - one that the PR doesn't open (now fixed), but also that the annotation is incorrect.  I think the intent of this issue, that it does it support GHE, isn't resolved yet, as the opened annotation won't work to tie a GHE repo into the catalog?

Specifically, it seems to grab the domain part of the defined GHE host and uses it, instead of the proper project slug `github.com/project-slug`: 
https://backstage.io/docs/features/software-catalog/well-known-annotations#githubcomproject-slug 

e.g.,

![image](https://user-images.githubusercontent.com/33203301/102114243-f9b4c900-3e07-11eb-9b46-fcdbdd8f1747.png)

Results in:

![image](https://user-images.githubusercontent.com/33203301/102114310-09cca880-3e08-11eb-8657-00df3f10fd6d.png)

**Comment by O5ten** [comment]
Yup, can totally reproduce that annotation thing. Was about to report it as another issue, but now i wont :). But sounds very close to the error i had in another place last week.

**Comment by adamdmharvey** [comment]
Here's a hint I think.  Looks like the TODO is already listed:

https://github.com/backstage/backstage/blob/257619c1ee9c49454fc32b6d7d2b07283ea8d738/plugins/catalog-backend/src/ingestion/LocationAnalyzer.ts#L40-L41

**Comment by adamdmharvey** [comment]
This may be just an easy one, since the annotation is ALWAYS static?  Just replacing `${source} with `github.com`???

**Comment by freben** [comment]
Yep, that code took a shortcut to easily support a couple of the most common public providers. It will need to be updated to instead read from the `integrations` config (with the help of the `packages/integration` package) to find what provider that the URL actually pointed to.

However, even if this annotation got the right key, as in for example `github.com/project-slug`, one problem remains - we haven't yet set in stone a format for the _value_, which lets you discern between various providers. If you for example have integrations both toward public github and an internal github enterprise, there isn't a well known format for encoding which one you mean in this annotation.

**Comment by adamdmharvey** [comment]
My team might actually have that use case.  e.g., we have our fair share of assets over in https://github.com/dxc-technology (like our https://github.com/dxc-technology/online-devops-dojo which you may have seen at DOES or AllDayDevOps this year 😁 ), with the bulk in our internal GHE.

In those cases, maintaining a slug for `github.com/project-slug: org/repo` in parallel say to `github-enterprise: org/repo` might be good? (would be an even further edge case to have multiple GHE instance allowances, though I will say we do fall into that scenario due to M&A activities, it's not one we would likely need to support so I'd be ok without it!!!)

**Comment by freben** [comment]
If you can easily hold off without ill effects then maybe that's safer. I gotta say that it's tempting to define this just as either the shorthand slug (assuming you have one provider, or it picks the "top" one), or the full URL to the repo (e.g. `https://github.com/dxc-technology/online-devops-dojo`) if one needs to discern between several. But again ... not defined quite yet and consumers will then need a facility to reliably parse what they need out of it. Which will be available when the `packages/integration` package is fleshed out later.

**Comment by adamdmharvey** [comment]
For now, I'm good with keeping just the raw `github.com/project-slug`, as we're only tracking one GitHub instance (not really dotcom yet), and have the multiple instances for a later enhancement.

Given that, this feature would need to know when to statically define the slug per provider so that the annotation is correct when opening the PR still.  Does that make sense to others?

**Comment by adamdmharvey** [comment]
Just to report, with the latest catalog-import, it still doesn't properly support GHE instances with the PRs.  The github slug it creates uses the wrong domain name - my corporate one rather than github.com - and thus the resulting entity doesn't work.

![image](https://user-images.githubusercontent.com/33203301/108914307-23a11d80-75f9-11eb-9f16-279b52649efe.png)

**Comment by freben** [comment]
@adamdmharvey  wait, is the KEY of the annotation wrong? That's unexpected; that bit should be hardcoded no matter what shouldn't it...

**Comment by adamdmharvey** [comment]
Ah you're spot on; sorry let me clarify.  The mechanics of actually OPENING the PR work; I correctly get a valid PR in my GHE repo.

But the visual display preview of the YAML in the sample card is wrong, as per the screen shot above.

**Comment by kuangp** [comment]
Hi, was wondering if there are any updates regarding this? We're also running into this issue with our GHE as we're starting to have projects to adopt Backstage.

**Comment by freben** [comment]
Sorry, the situation is the same. Until it's resolved, we recommend registering full YAML urls instead of directories. That flow should work perfectly for your GHE.

**Comment by kuangp** [comment]
I have a [PR](https://github.com/backstage/backstage/pull/5444) waiting for review to address this
